addons/purchase/purchase.py

```diff
@@ -1190,41 +1182,40 @@ def onchange_product_id(self, cr, uid, ids, pricelist_id, product_id, qty, uom_i
         # - determine product_qty and date_planned based on seller info
         if not date_order:
             date_order = fields.datetime.now()
-
-
+        date = time.strftime(DEFAULT_SERVER_DATE_FORMAT)
         supplierinfo = False
         precision = self.pool.get('decimal.precision').precision_get(cr, uid, 'Product Unit of Measure')
-        for supplier in product.seller_ids:
-            if partner_id and (supplier.name.id == partner_id):
-                supplierinfo = supplier
+        cr.execute("SELECT s.id \
+                    FROM product_supplierinfo AS s \
+                    WHERE (s.name = %s) AND (s.product_tmpl_id = %s) \
+                    AND ((s.date_start IS NULL OR s.date_start<=%s) AND (s.date_end IS NULL OR s.date_end>=%s)) ORDER BY s.min_qty ,s.price asc ", (partner_id, product.product_tmpl_id.id, date, date))
```

Comment:
fetch based on s.price asc ?  by default it's ascending order no need to define asc.